### PR TITLE
[libc][NFC] Simplify FloatProperties implementation

### PR DESCRIPTION
**Pull request by gchatelet**
This is a follow up on #74481 that migrates code from all specializations into `FPCommonProperties` (except for `EXPLICIT_BIT_MASK` in the `X86_Binary80` specialization)

### CODE REVIEW

**Comment by gchatelet** [comment]
The interesting change is https://github.com/llvm/llvm-project/pull/74821/commits/a416d0b2bea55d6ec4ae90f4c75eabafa2fb310b.

File: libc/src/__support/FPUtil/FloatProperties.h [hunk 1]
```diff
@@ -80,27 +80,96 @@ template <> struct FPBaseProperties<FPType::X86_Binary80> {
...
+      mask_trailing_ones<UIntType, SIGN_BITS>() << SIGN_MASK_SHIFT;
+  LIBC_INLINE_VAR static constexpr UIntType FP_MASK =
+      mask_trailing_ones<UIntType, TOTAL_BITS>();
+  static_assert((SIG_MASK & EXP_MASK & SIGN_MASK_) == 0, "masks disjoint");
+  static_assert((SIG_MASK | EXP_MASK | SIGN_MASK_) == FP_MASK, "masks covers");
+
+  //
+  LIBC_INLINE_VAR static constexpr int FRACTION_BITS =
+      SIG_BITS - EXPLICIT_SIG_BITS;
+  LIBC_INLINE_VAR static constexpr UIntType QNAN_MASK = UIntType(1)
```
[the remark this answers is not in the document]
Comment (gchatelet) [reply libc/src/__support/FPUtil/FloatProperties.h:142]: I made a quick bit printer https://godbolt.org/z/TGqxsq1KM and I think I've messed up indeed. That said tests are green... 😅

File: libc/src/__support/FPUtil/FloatProperties.h [hunk 2]
```diff
@@ -80,27 +80,96 @@ template <> struct FPBaseProperties<FPType::X86_Binary80> {
...
+  using BitsType = UIntType;
+
+  LIBC_INLINE_VAR static constexpr uint32_t BIT_WIDTH = TOTAL_BITS;
+  LIBC_INLINE_VAR static constexpr uint32_t MANTISSA_WIDTH = FRACTION_BITS;
+  LIBC_INLINE_VAR static constexpr uint32_t MANTISSA_PRECISION =
+      MANTISSA_WIDTH + 1;
+  LIBC_INLINE_VAR static constexpr BitsType MANTISSA_MASK =
+      (BitsType(1) << MANTISSA_WIDTH) - 1;
+  LIBC_INLINE_VAR static constexpr uint32_t EXPONENT_WIDTH = EXP_BITS;
+  LIBC_INLINE_VAR static constexpr uint32_t EXPONENT_BIAS = EXP_BIAS;
```
[the remark this answers is not in the document]
Comment (gchatelet) [reply libc/src/__support/FPUtil/FloatProperties.h]: I did a quick run through the code base and indeed most use of `EXPONENT_BIAS` involves signed arithmetic so it seems favorable to change its type to `int32_t` in the long run - not in the patch though.

File: libc/src/__support/FPUtil/FloatProperties.h [hunk 3]
```diff
@@ -27,116 +27,173 @@ enum class FPType {
...
+  LIBC_INLINE_VAR static constexpr int SIG_MASK_SHIFT = 0;
+  LIBC_INLINE_VAR static constexpr int EXP_MASK_SHIFT = SIG_BITS;
+  LIBC_INLINE_VAR static constexpr int SIGN_MASK_SHIFT = SIG_BITS + EXP_BITS;
+
+  // Masks
+  LIBC_INLINE_VAR static constexpr UIntType SIG_MASK =
+      mask_trailing_ones<UIntType, SIG_BITS>() << SIG_MASK_SHIFT;
+  LIBC_INLINE_VAR static constexpr UIntType EXP_MASK =
+      mask_trailing_ones<UIntType, EXP_BITS>() << EXP_MASK_SHIFT;
+  LIBC_INLINE_VAR static constexpr UIntType SIGN_MASK_ =
```
Comment (lntue) [inline libc/src/__support/FPUtil/FloatProperties.h:127]: nit: it's a little odd to have extra `_` trailing with `SIGN_MASK`, but not with other `*_MASK`.

File: libc/src/__support/FPUtil/FloatProperties.h [hunk 4]
```diff
@@ -27,116 +27,173 @@ enum class FPType {
...
+    return UIntType(1) << position;
+  }
 
-  static constexpr BitsType EXP_MANT_MASK =
-      MANTISSA_MASK | EXPLICIT_BIT_MASK | EXPONENT_MASK;
-  static_assert(EXP_MANT_MASK == (~SIGN_MASK & FULL_WIDTH_MASK),
-                "Exponent and mantissa masks are not as expected.");
+  LIBC_INLINE_VAR static constexpr UIntType QNAN_MASK =
+      UP::ENCODING == FPEncoding::X86_ExtendedPrecision
+          ? bit_at(SIG_BITS - 1) | bit_at(SIG_BITS - 2) // 0b1100...
```
Comment (lntue) [inline libc/src/__support/FPUtil/FloatProperties.h:140]: Is there a reason for not including the exponent field to `(Q|S)NAN_MASK`?

File: libc/src/__support/FPUtil/FloatProperties.h (hunk 3 above)
Comment (gchatelet) [reply libc/src/__support/FPUtil/FloatProperties.h:127]: Added a comment to clarify.

File: libc/src/__support/FPUtil/FloatProperties.h (hunk 4 above)
Comment (gchatelet) [reply libc/src/__support/FPUtil/FloatProperties.h:140]: I don't quite know what to do with `(Q|S)NAN_MASK` in the future, the purpose of this patch is to be a NFC and to retain the original semantics of `QUIET_NAN_MASK`.
https://github.com/llvm/llvm-project/blob/75193b192ad92e69236930dc35c262786a95f472/libc/src/__support/FPUtil/FloatProperties.h#L138-L139

The current documentation says
https://github.com/llvm/llvm-project/blob/75193b192ad92e69236930dc35c262786a95f472/libc/src/__support/FPUtil/FloatProperties.h#L135-L137

So it seems that being a `NaN` is a prerequisite for using this mask. Now I don't mind adding the exponent mask as well if you feel like it. Let me know!

**Review by lntue — APPROVED** [review]

**Pull request merged**